### PR TITLE
[doc] Add architecture decision records

### PR DESCRIPTION
**Pull request by adangel**
## Describe the PR

This is a first draft on adding explicit decision records to our project.
It came up as a question in #3766 about Kotlin usage.

I'm not sure in what details we would need to describe the decisions or which decisions at all. I guess, we need to learn this over time.

Feel free to give feedback or ask questions - the goal is to have the decision records as an agreement and understandable by everybody.

I've integrated these into the doc site, they appear under "Project documentation":
grafik

But since these documents are simple markdown files, they can also be seen directly on github.

### CODE REVIEW

File: docs/pages/pmd/projectdocs/decisions.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+---
+title: Architecture Design Decisions
```
Comment (adangel) [inline docs/pages/pmd/projectdocs/decisions.md:2]:
```suggestion
title: Architecture Decisions
```

File: docs/pages/pmd/projectdocs/decisions/adr-1.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+{{ page.adr_status }}
+
+# Consequences
+
+Explicitly documenting decisions has the benefit that new developers joining the projects know about the decisions
+and can read the context and consequences of the decisions. This will likely also improve the overall quality
+as the decisions need to be formulated and written down. Everybody is on the same page.
+
+However, this also adds additional tasks, and it takes time to write down and document the decisions.
```
Comment (adangel) [inline docs/pages/pmd/projectdocs/decisions/adr-1.md:52]: More documentation that might become old, stale if it is not kept up to date.

File: docs/pages/pmd/projectdocs/decisions/adr-1.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+# Decision
+
+We will document the decisions we make as a project as a collection of "Architecture Decision Records".
+In order to keep it simple, we will use only a simple template proposed by Michael Nygard.
+The documents are stored together with the source code and are part of the generated documentation site.
+
+# Status
+
+{{ page.adr_status }}
```
Comment (adangel) [inline docs/pages/pmd/projectdocs/decisions/adr-1.md:44]: Maybe we should also add a date, when this decision was accepted.

File: docs/pages/pmd/projectdocs/decisions/adr-1.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+by Michael Nygard.
+
+There are many templates around to choose from. <https://github.com/joelparkerhenderson/architecture-decision-record>
+gives a nice summary.
+
+# Decision
+
+We will document the decisions we make as a project as a collection of "Architecture Decision Records".
+In order to keep it simple, we will use only a simple template proposed by Michael Nygard.
+The documents are stored together with the source code and are part of the generated documentation site.
```
Comment (adangel) [inline docs/pages/pmd/projectdocs/decisions/adr-1.md:40]: Another addition: When we update those decision records, we might add more info in the context section - we should consider add new info always with a date stamp to make it easy. The documents would then be more like living documents.

File: docs/pages/pmd/projectdocs/decisions/adr-1.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+Past decisions have been made within context and the context can change. Therefore, past decisions can still be
+valid today, or they don't apply anymore. In that case, the decision should be revisited.
+
+See also the blog post [Documenting Architecture Decisions](https://cognitect.com/blog/2011/11/15/documenting-architecture-decisions)
+by Michael Nygard.
+
+There are many templates around to choose from. <https://github.com/joelparkerhenderson/architecture-decision-record>
+gives a nice summary.
+
```
Comment (adangel) [inline docs/pages/pmd/projectdocs/decisions/adr-1.md:35]: There's also https://adr.github.io/

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">No regression tested rules have been changed.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: docs/pages/pmd/projectdocs/decisions/adr-1.md (hunk 4 above)
Comment (adangel) [reply docs/pages/pmd/projectdocs/decisions/adr-1.md:40]: Or maybe even adding a "Change history" section? Maybe too much...

File: docs/pages/pmd/projectdocs/decisions/adr-2.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
+---
+title: ADR 2 - Policy on the use of Kotlin for development
```
Comment (adangel) [inline docs/pages/pmd/projectdocs/decisions/adr-2.md:2]: Maybe we should not make this specifically about Kotlin, but about non-Java languages in general? Or keep it for now limited to Kotlin? Then my example with Scala should be removed...

File: docs/pages/pmd/projectdocs/decisions/adr-1.md (hunk 3 above)
Comment (oowekyala) [reply docs/pages/pmd/projectdocs/decisions/adr-1.md:44]: Yes absolutely

File: docs/pages/pmd/projectdocs/decisions/adr-1.md (hunk 2 above)
Comment (oowekyala) [reply docs/pages/pmd/projectdocs/decisions/adr-1.md:52]: I don't think it necessarily needs to be up to date, as long as the date of the decision is prominently displayed somewhere.

**Comment by oowekyala** [comment]
This is a great idea!

I have a few questions/ topics for general discussion. I'll add my comments to the kotlin ADR as review comments
- What is the process to accept an ADR? In this PR the new ADRs have status "proposed". Is maintainer consensus enough to make them "accepted"?
- What's the process to challenge or amend an ADR? Open a GH issue maybe? A PR editing the ADR?

I think we should document these processes in ADR 0... Or at least sketch the expected "lifecycle" of an ADR, with different statuses

I'm wondering about the scope of ADRs. Do "architecture decisions" include eg proposals for large features (like eg JEPs)? Is there a difference with RFC issues? Maybe issues are more implementation related.

Would the following be appropriate topics for ADRs, or am I missing the point:
- policy on rule contributions (criteria for inclusion and quality standards)
- policy on language module contributions (same)
- binary compatibility guarantees (including describing internal packages and `@InternalApi`)
- code of conduct

File: docs/pages/pmd/projectdocs/decisions/adr-2.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+* Additionally: from now on, we allow to have the individual language modules be implemented in different languages
+  when it makes sense. So, a language module might decide to use plain Java (like now) or also Kotlin
+  (or other languages if it fits).
+* When mixing languages (e.g. Java + Kotlin), we need to care that the modules can still be used with plain Java.
+  E.g. when writing custom rules: `pmd-java` provides a couple of APIs for rules (like symbol table, type resolution)
+  and we should not force the users to use Kotlin (at least not for language modules which already exist and
+  for which users might have written custom rules in Java already).
+* It is also possible to write the entire language module in Kotlin only. Then the rules would be written in Kotlin
+  as well. And the possible problems when mixing languages are gone. But that applies only for new language modules.
+  For compatibility reasons an existing language modules shouldn't be rewritten into Kotlin. That would be a
```
Comment (oowekyala) [inline docs/pages/pmd/projectdocs/decisions/adr-2.md:50]: > For compatibility reasons an existing language modules shouldn't be rewritten into Kotlin.

I think this is too strong a statement. They shouldn't be rewritten *incompatibly* to Kotlin, ok.

I think if converting a class to kotlin is possible in a binary and source compatible way, then it should not be forbidden.
By extension, converting a number of classes, or a whole package, or a whole language module, should not be forbidden either, provided we respect our own compatibility guarantees.

File: docs/pages/pmd/projectdocs/decisions/adr-2.md (hunk 6 above)
Comment (oowekyala) [reply docs/pages/pmd/projectdocs/decisions/adr-2.md:2]: I think this document is well-written and I like the amount of context specifically on kotlin. I would say, we should keep it as is

File: docs/pages/pmd/projectdocs/decisions/adr-1.md (hunk 2 above)
Comment (adangel) [reply docs/pages/pmd/projectdocs/decisions/adr-1.md:52]: I'll add the "last_updated" info to the status section.

File: docs/pages/pmd/projectdocs/decisions/adr-1.md (hunk 3 above)
Comment (adangel) [reply docs/pages/pmd/projectdocs/decisions/adr-1.md:44]: Let's keep for now the "last_updated" date and see how it works. Maybe we don't need a "adr_date_accepted" field or "adr_valid_from" etc.

File: docs/pages/pmd/projectdocs/decisions/adr-1.md (hunk 4 above)
Comment (adangel) [reply docs/pages/pmd/projectdocs/decisions/adr-1.md:40]: Added section "Change History" to the template.

File: docs/pages/pmd/projectdocs/decisions/adr-2.md (hunk 6 above)
Comment (adangel) [reply docs/pages/pmd/projectdocs/decisions/adr-2.md:2]: :+1:

File: docs/pages/pmd/projectdocs/decisions/adr-2.md (hunk 7 above)
Comment (adangel) [reply docs/pages/pmd/projectdocs/decisions/adr-2.md]: I've changed the part to be now:

* It is also possible to write the entire language module in Kotlin only. Then the rules would be written in Kotlin
  as well. And the possible problems when mixing languages are gone. But that applies only for new language modules.
* When refactoring an existing language module from Java only to introduce Kotlin, care needs to be taken to
  not make incompatible changes. If compatibility (binary or source) can't be maintained, then that would be a
  major version change.

**Comment by adangel** [comment]
> This is a great idea!
> 
> I have a few questions/ topics for general discussion. I'll add my comments to the kotlin ADR as review comments
>
> * What is the process to accept an ADR? In this PR the new ADRs have status "proposed". Is maintainer consensus enough to make them "accepted"?
> * What's the process to challenge or amend an ADR? Open a GH issue maybe? A PR editing the ADR?
> 
> I think we should document these processes in ADR 0... Or at least sketch the expected "lifecycle" of an ADR, with different statuses
> 

Good points - I've added a short description what I think could work in the ADR-1. Basically we use PRs to propose new ADRs and modify existing ADRs.
I've not sketched a complete lifecycle, though, just a short description in the "Decision" section.

> I'm wondering about the scope of ADRs. Do "architecture decisions" include eg proposals for large features (like eg JEPs)? Is there a difference with RFC issues? Maybe issues are more implementation related.

Hm... I don't think, that large features should be written as ADRs. A feature sounds too specific for me. Maybe if we can come up with a general direction of what PMD should be or shouldn't be, that might be an ADR. E.g. (don't take these examples serious... :smile: ) "PMD will concentrate the work on the core functions needed for static analysis. Any integration into plugins, tools, etc. is out of scope for the project" or "PMD won't use GitHub anymore, because it is not open source" or "PMD will only support languages where free tooling is available" or "PMD will drop support for Windows".

RFC issues... I could imagine, that an ADR could originate from an RFC issue. So that the discussion on the issue leads us to writing the result down as an ADR.

If issues are implementation related, that could be too specific for an ADR.

> 
> Would the following be appropriate topics for ADRs, or am I missing the point:
>
> * policy on rule contributions (criteria for inclusion and quality standards)
> * policy on language module contributions (same)
> * binary compatibility guarantees (including describing internal packages and `@InternalApi`)
> * code of conduct

I think, these are all good examples for ADRs. I'd say it's a bit of a mixture of architecture and processes (policies).

File: docs/pages/pmd/projectdocs/decisions/adr-1.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
+---
+title: ADR 1 - Use architecture decision records
+sidebar: pmd_sidebar
+permalink: pmd_projectdocs_decisions_adr_1.html
+sidebaractiveurl: /pmd_projectdocs_decisions.html
+adr: true
+# Proposed / Accepted / Deprecated / Superseded
+adr_status: "Proposed"
```
Comment (adangel) [inline docs/pages/pmd/projectdocs/decisions/adr-1.md:8]:
```suggestion
adr_status: "Accepted"
```

File: docs/pages/pmd/projectdocs/decisions/adr-1.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+Explicitly documenting decisions has the benefit that new developers joining the projects know about the decisions
+and can read the context and consequences of the decisions. This will likely also improve the overall quality
+as the decisions need to be formulated and written down. Everybody is on the same page.
+
+However, this also adds additional tasks, and it takes time to write down and document the decisions.
+
+# Change History
+
+2022-09-06: Added section "Change History" to the template. Added "Last updated" to "Status" section.
```
Comment (adangel) [inline docs/pages/pmd/projectdocs/decisions/adr-1.md:66]:
```suggestion
2022-09-30: Status changed to "Accepted".

2022-09-06: Added section "Change History" to the template. Added "Last updated" to "Status" section.
```

File: docs/pages/pmd/projectdocs/decisions/adr-2.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
+---
+title: ADR 2 - Policy on the use of Kotlin for development
+sidebar: pmd_sidebar
+permalink: pmd_projectdocs_decisions_adr_2.html
+sidebaractiveurl: /pmd_projectdocs_decisions.html
+adr: true
+# Proposed / Accepted / Deprecated / Superseded
+adr_status: "Proposed"
```
Comment (adangel) [inline docs/pages/pmd/projectdocs/decisions/adr-2.md:8]:
```suggestion
adr_status: "Accepted"
```

File: docs/pages/pmd/projectdocs/decisions/adr-2.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+In the future we might also consider to use other languages than Kotlin, e.g. for `pmd-scala` Scala might make sense.
+
+On the other side, other IDEs than IntelliJ IDEA will have a difficult time to deal with PMD's source code
+when Kotlin is used. Eclipse can't be used practically anymore.
+
+Maintaining a polyglot code base with multiple languages is likely to be more challenging.
+
+# Change History
+
+2022-07-28: Proposed initial version.
```
Comment (adangel) [inline docs/pages/pmd/projectdocs/decisions/adr-2.md:69]:
```suggestion
2022-09-30: Changed status to "Accepted".

2022-07-28: Proposed initial version.
```

**Review by adangel — COMMENTED** [review]
I'll proceed now and mark our first two ADRs as "Accepted".

Keep in mind, we can still change these decisions - they are not cut in stone.

**Pull request merged**